### DISCUSSION TITLE
Question on Specifying the Target Field

### DISCUSSION THREAD

**Discussion by jamie2minchae** [bfieldtools/bfieldtools#59] [Q&A] [answered]
Thank you for creating such wonderful software. I have a question on specifying the target field.

 For us, only the z-component of the target magnetic field matters and it is not necessary to specify the (x, y)-components.

 However, I think that the (x, y, z)-components of the target field should be specified simultaneously, which will likely impose extra constraints on coil designs. Note that setting (x, y)=(0, 0) is an explicit specification of the target field, and we would rather want the coil design not to care about the generated (x, y)-fields, which can potentially have arbitrary values.

Can you help us on implementing this in your bfieldtools software package?

Perhaps the "optimize_streamfunction" can only use the z-component of the target field to generate optimal coil designs, or the weight of the (x, y)-target-field constraints can be tuned down within the software. However, we are not sure on how to implement this. We are using the object "minimum_Ohmic_power".

Your fast response will be very much appreciated. Thank you in advance.

**Comment by Astrei** [comment]
Maybe the abs_error parameter can help with it somehow? It looks like it's an array for the different field components, but I am not sure.

**Comment by rzetter** [comment, marked as answer]
Hi! This is actually very easy to do. Taking a snippet of code from the biplanar_coil_design.py example, you would specify the optimization spec as below. I've simply chosen the component I care about (In your case, Z i.e. component 2) in the B_coupling matrix as well as the target field. The dimensions of the B_coupling matrix correspond to (number of field points, number of field components, number of mesh points/basis functions).


    target_field = np.zeros(target_points.shape)
    target_field[:, 2] += 1
    
    target_spec = {
        "coupling": coil.B_coupling(target_points)[:,2, :],
        "abs_error": 0.01,
        "target": target_field[:,2],
    }
    
    bfield_specification = [target_spec]

**Comment by jamie2minchae** [comment]
Thank you very much!!
